private/protocol/restjson/unmarshal_error_test.go

```diff
@@ -7,13 +7,169 @@ import (
 	"encoding/hex"
 	"io/ioutil"
 	"net/http"
+	"reflect"
 	"strings"
 	"testing"
 
+	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/aws/awserr"
 	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/aws/aws-sdk-go/private/protocol"
 )
 
+const unknownErrJSON = `{"code":"UnknownError", "message":"error message", "something":123}`
+const simpleErrJSON = `{"code":"SimpleError", "message":"some message", "foo":123}`
+
+type SimpleError struct {
+	_ struct{} `type:"structure"`
+	error
+
+	Message2 *string `type:"string" locationName:"message"`
+	Foo      *int64  `type:"integer" locationName:"foo"`
+}
+
+const otherErrJSON = `{"code":"OtherError", "message":"some message"}`
+const complexCodeErrJSON = `{"code":"OtherError:foo:bar", "message":"some message"}`
+
+type OtherError struct {
+	_ struct{} `type:"structure"`
+	error
+
+	Message2 *string `type:"string" locationName:"message"`
+}
+
+const complexErrJSON = `{"code":"ComplexError", "message":"some message", "foo": {"bar":"abc123", "baz":123}}`
+
+type ComplexError struct {
+	_ struct{} `type:"structure"`
+	error
+
+	Message2  *string      `type:"string" locationName:"message"`
+	Foo       *ErrorNested `type:"structure" locationName:"foo"`
+	HeaderVal *string      `type:"string" location:"header" locationName:"some-header"`
+	Status    *int64       `type:"integer" location:"statusCode"`
+}
+type ErrorNested struct {
+	_ struct{} `type:"structure"`
+
+	Bar *string `type:"string" locationName:"bar"`
+	Baz *int64  `type:"integer" locationName:"baz"`
+}
+
+func TestUnmarshalTypedError(t *testing.T) {
```

Comment:
We should include test cases for verifying proper behavior when the `x-amzn-error*` headers are present